trie/iterator.go

```diff
@@ -836,3 +836,41 @@ func (it *unionIterator) Error() error {
 	}
 	return nil
 }
+
+// prefixIterator is a wrapper around NodeIterator that stops iteration
+// when it leaves a subtree with a specific prefix.
+type prefixIterator struct {
+	NodeIterator
+	prefix []byte
```

Comment:
You don't just need a prefix, you also need a stopping point. The issue is about iterating a subtree, not just starting at a random point.